### PR TITLE
Added TableManager. Check for HBase table existence during table creation

### LINKED ISSUE #58: Check HBase table existence when seggregatedTable is set
Right now, foxtrot takes a leap of faith and assumes that the table is present. It looks like a good idea to do this check.

However, there will be a severe performance impact of doing this. We might have to be smart about this and keep refreshing the table metadata in the background to all foxtrot nodes.

### PR DESCRIPTION
**Pull request by r0goyal**
closes #58

### CODE REVIEW

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public void delete(String tableName) throws TableManagerException {
+        // TODO Implement this once downstream implications are figured out
+    }
+
+    private void validateTableParams(Table table) throws TableManagerException {
+        if (table == null || table.getName() == null || table.getName().trim().isEmpty() || table.getTtl() <= 0) {
+            throw new TableManagerException(TableManagerException.ErrorCode.BAD_REQUEST, "Invalid Table Params");
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java]: Most of these conditions are impossible due to javax validations.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            throw new TableManagerException(TableManagerException.ErrorCode.INTERNAL_ERROR, e);
+        }
+    }
+
+    @Override
+    public Table get(String name) throws TableManagerException {
+        try {
+            Table table = metadataManager.get(name);
+            if (table == null) {
+                throw new TableManagerException(TableManagerException.ErrorCode.TABLE_NOT_FOUND, "table_does_not_exists");
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java]: exist .. and why the underscores? Also, add the tablename for which this error is coming.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            }
+            return table;
+        } catch (TableManagerException e) {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java]: unnecessary

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public void update(Table table) throws TableManagerException {
+        validateTableParams(table);
+        try {
+            if (!metadataManager.exists(table.getName())) {
+                throw new TableManagerException(TableManagerException.ErrorCode.TABLE_NOT_FOUND, "table_does_not_exists");
+            }
+            metadataManager.save(table);
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java]: Call dummy functions...

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/QueryStore.java [hunk 5]
```diff
@@ -26,21 +27,23 @@
  * Date: 12/03/14
  * Time: 9:25 PM
  */
-public interface QueryStore {
+public interface QueryStore extends Identifiable {
 
-    public void save(final String table, final Document document) throws QueryStoreException;
+    void init(final String table) throws QueryStoreException;
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/QueryStore.java]: call this createTable() or initializeTable() to make the intent explicit.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/impl/hbase/HBaseDataStore.java [hunk 6]
```diff
@@ -56,6 +56,16 @@ public HBaseDataStore(HbaseTableConnection tableWrapper, ObjectMapper mapper) {
     }
 
     @Override
+    public void init(Table table) throws DataStoreException {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/impl/hbase/HBaseDataStore.java]: call this createTable() or initializeTable() to make the intent explicit.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Identifiable.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+/**
+ * Created by rishabh.goyal on 05/12/15.
+ */
+
+public interface Identifiable {
+
+    String getId();
+
+}
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Identifiable.java]: This class and the function it generates is useless.

File: foxtrot-core/pom.xml [hunk 8]
```diff
@@ -80,7 +79,7 @@
         <dependency>
             <groupId>com.flipkart.foxtrot</groupId>
             <artifactId>foxtrot-common</artifactId>
-            <version>0.3</version>
+            <version>${project.parent.version}</version>
```
Comment (santanusinha) [inline foxtrot-core/pom.xml]: Good .. Can you also move all dependency versions to properties? Can be a separate pull request.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Action.java [hunk 9]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright 2014 Flipkart Internet Pvt. Ltd.
- *
+ * <p/>
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Action.java]: Remove all such lines. No HTML formatting in source code. Atleast not in the license section.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/DataStore.java [hunk 10]
```diff
@@ -25,9 +26,15 @@
  * Date: 12/03/14
  * Time: 9:17 PM
  */
-public interface DataStore {
+public interface DataStore extends Identifiable {
+
+    public void init(final Table table) throws DataStoreException;
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/DataStore.java]: call this createTable() or initializeTable() to make the intent explicit.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/DataStore.java [hunk 11]
```diff
@@ -25,9 +26,15 @@
  * Date: 12/03/14
  * Time: 9:17 PM
  */
-public interface DataStore {
+public interface DataStore extends Identifiable {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/DataStore.java]: Inherting this interface is not required as the getId function is not used anywhere.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/QueryStore.java [hunk 12]
```diff
@@ -26,21 +27,23 @@
  * Date: 12/03/14
  * Time: 9:25 PM
  */
-public interface QueryStore {
+public interface QueryStore extends Identifiable {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/QueryStore.java]: Remove the inheritance

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchQueryStore.java [hunk 13]
```diff
@@ -325,4 +330,9 @@ public void cleanup(Set<String> tables) throws QueryStoreException {
                     String.format("Unable to delete Indexes - %s", indicesToDelete), ex);
         }
     }
+
+    @Override
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchQueryStore.java]: Remove

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/TableManager.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package com.flipkart.foxtrot.core.table;
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/TableManager.java]: License .. this is not required now, but before release we need to ensure all source files have license.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+                throw new TableManagerException(TableManagerException.ErrorCode.BAD_REQUEST, "table already exists");
+            }
+            queryStore.init(table.getName());
+            dataStore.init(table);
+            metadataManager.save(table);
+        } catch (DataStoreException e) {
+            switch (e.getErrorCode()) {
+                case TABLE_NOT_FOUND:
+                    throw new TableManagerException(TableManagerException.ErrorCode.BAD_REQUEST, e);
+                case STORE_CONNECTION:
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java]: This and the next case can be merged.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java (hunk 1 above)
Comment (r0goyal) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java]: Javax validation only works when using HTTP API today. It can later be used if we allow table creation using FQL

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Identifiable.java (hunk 7 above)
Comment (r0goyal) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Identifiable.java]: Was using it in the interim. Will remove it

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java (hunk 1 above)
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/FoxtrotTableManager.java]: We still need to pass it through the validator itself in that case .. otherwise there will be two sources of checks.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/TableManager.java (hunk 14 above)
Comment (r0goyal) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/TableManager.java]: Will add license to all files in one go later

**Comment by santanusinha** [comment]
Looks good overall .. minor quibbles.
